[sparsity][refactor] Changing linear row/col control

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #60728 [quant][sparsity] Generic convert function
* #59771 [sparsity] Lambda Scheduler
* #59770 [sparsity] Base sparsity level scheduler class
* #58955 [sparsity] WeightNormSparsifier
* #58707 [sparsity][refactor] Import factoring out
* #58704 [sparsity] Sparsifier class
* #58705 [sparsity] Sparsity parametrization
* **#60850 [sparsity][refactor] Changing linear row/col control**
* #60887 [sparsity] Add sparsity tests to run_test.py

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29465900](https://our.internmc.facebook.com/intern/diff/D29465900)

Commit:
[sparsity][refactor] Changing linear row/col control

[ghstack-poisoned]

Commit:
Update on "[sparsity][refactor] Changing linear row/col control"




[ghstack-poisoned]

Commit:
Update on "[sparsity][refactor] Changing linear row/col control"




[ghstack-poisoned]

Commit:
Update on "[sparsity][refactor] Changing linear row/col control"


Test Plan:

```
python test/test_ao_sparsity.py
```

[ghstack-poisoned]

raghuramank100 commented on torch/ao/nn/sparse/quantized/utils.py:
```diff
@@ -19,20 +19,20 @@ class QNNPACKLinearBlockSparsePattern:
-        QNNPACKLinearBlockSparsePattern.prev_col_block_size = QNNPACKLinearBlockSparsePattern.col_block_size
-        QNNPACKLinearBlockSparsePattern.row_block_size = row_block_size
-        QNNPACKLinearBlockSparsePattern.col_block_size = col_block_size
+        LinearBlockSparsePattern.rlock.acquire()
```
Why is there an rlock acquire here? From point of view of modularity, should this be in C++ as part of the kernels? Do we need this for both FBGEMM and QNNPACK?

z-a-f replied:
This is a lock for thread safety on python level.

raghuramank100 replied:
cc @vkuzo  Wondering if there are any other side effects to this?

vkuzo replied:
This PR is just renaming things, the code was added by @kimishpatel and is not really related to this PR afaik.  @kimishpatel , would you mind chiming in with some more context on why this is here?

Agreed that it's unexpected to see this in a sparsity utils file, ideally this stuff would be in a lower level multithreading utility somewhere.  If it is needed here, can we add a comment describing why?

z-a-f replied:
I feel like I was explaining this for some time already. Singletons in python are not thread-safe. To make them thread safe, it's a common practice to add locks in there. Adding explanation on that is fine, which I could do in a separate PR. But I am still confused about why it's blocking landing of the current refactoring PR.

vkuzo replied:
> But I am still confused about why it's blocking landing of the current refactoring PR.

+1, I was only responding because Raghu tagged me, but this discussion is not related to the goal of this PR.

kimishpatel replied:
@vkuzo @z-a-f @raghuramank100, originally I added this as an C++ equivalent RAII object for context management. Context being managed here is the block pattern of sparsity. Reason we need context management: At the time of this code writing, we did not have sparsity workflow (as mentioned in the comment) and we were piggy-backing on quantization workflow. Since quantization configs dont have a way to specify sparsity pattern I added this context manager that said "now sparsity pattern is *this* and for all SparseDynamicLinearPackedParam that you create, use the specified sparsity captured by the context". This was need because QNNPACK supported 1x4 and 8x1 unliked fbgemm that supported only 1x4. So we needed a way to specify which sparsity pattern do we want during convert phase of the sparsity workflow. Let me know if this adds the needed context or made it worse?

Once we have sparsity workflow whose config captures sparsity pattern we should not need this.

z-a-f replied:
@raghuramank100 Should we add comments in file to avoid confusion? As we discussed in person this is a temporary measure until we are done (the same as @kimishpatel mentioned), so I think we don't need to explain it. However, it might be useful for posterity and paper-trail.

Either way, if there are no other concerns, I would appreciate a stamp :) If comments needed, I will send it as a separate PR

z-a-f replied:
> > But I am still confused about why it's blocking landing of the current refactoring PR.
> 
> +1, I was only responding because Raghu tagged me, but this discussion is not related to the goal of this PR.

Oh I didn't mean to sound annoyed -- I welcome your contribution (and request it for other PRs as well). I just got surprised that this blew bigger than I thought it would :)

raghuramank100 reviewed (changes requested):
Please see comments

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29465900).

Commit:
Update on "[sparsity][refactor] Changing linear row/col control"


Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29465900](https://our.internmc.facebook.com/intern/diff/D29465900)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29465900).

raghuramank100 commented on test/ao/sparsity/test_kernels.py:
```diff
@@ -10,10 +10,9 @@
 from torch import nn
 from torch.ao.nn.sparse import quantized as ao_nn_sq
-from torch.ao.nn.sparse.quantized.utils import QNNPACKLinearBlockSparsePattern
+from torch.ao.nn.sparse.quantized.utils import LinearBlockSparsePattern
```
Please add a TODO: In a later PR where we modify from_float to take the sparsity pattern as an argument and this pattern will not be needed.

Commit:
Update on "[sparsity][refactor] Changing linear row/col control"


Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29465900](https://our.internmc.facebook.com/intern/diff/D29465900)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29465900).

Commit:
Update on "[sparsity][refactor] Changing linear row/col control"


Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29465900](https://our.internmc.facebook.com/intern/diff/D29465900)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29465900).
